File: value/src/main/java/com/bnorm/barkeep/model/BarValue.java
// Copyright 2016 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.model;

import com.google.auto.value.AutoValue;
import com.google.common.collect.ImmutableSet;

@AutoValue
public abstract class BarValue implements Bar {

  @Override
  public abstract ImmutableSet<Ingredient> getIngredients();

  public static Builder builder() {
    return new AutoValue_BarValue.Builder();
  }

  @AutoValue.Builder
  public abstract static class Builder {

    public abstract BarValue build();

    public abstract Builder setId(Long id);

    public abstract Builder setTitle(String title);

    public abstract Builder setDescription(String description);

    public abstract Builder setOwner(User owner);

    abstract ImmutableSet.Builder<Ingredient> ingredientsBuilder();

    public Builder addIngredient(Ingredient ingredient) {
      ingredientsBuilder().add(ingredient);
      return this;
    }
  }
}

File: model-value/src/main/kotlin/com/bnorm/barkeep/model/value/BarValue.kt
// Copyright 2017 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.model.value

import com.bnorm.barkeep.model.Bar
import com.bnorm.barkeep.model.BarSpec
import com.bnorm.barkeep.model.Ingredient
import com.bnorm.barkeep.model.User
import com.squareup.moshi.FromJson
import com.squareup.moshi.ToJson

data class BarSpecValue(override val title: String?,
                        override val description: String? = null,
                        override val owner: User? = null) : BarSpec

data class BarValue(override val id: Long = -1,
                    override val title: String?,
                    override val description: String? = null,
                    override val owner: User? = null,
                    override val ingredients: Set<Ingredient> = emptySet()) : Bar

object BarValueAdapter {
  @ToJson fun toJson(bar: BarSpec): BarSpecValue {
    return BarSpecValue(bar.title, bar.description, bar.owner)
  }

  @FromJson fun fromJson(json: BarSpecValue): BarSpec {
    return json
  }

  @ToJson fun toJson(bar: Bar): BarValue {
    return BarValue(bar.id, bar.title, bar.description, bar.owner, bar.ingredients)
  }

  @FromJson fun fromJson(json: BarValue): Bar {
    return json
  }
}
